addon/components/frost-autocomplete.js

```diff
@@ -0,0 +1,325 @@
+import Ember from 'ember'
+import computed, {readOnly} from 'ember-computed-decorators'
+import {task, timeout} from 'ember-concurrency'
+import {PropTypes} from 'ember-prop-types'
+import layout from '../templates/components/frost-autocomplete'
+import {keyCodes} from '../utils'
+import Component from './frost-component'
+
+const {get, isEmpty, on, run, typeOf} = Ember
+const {BACKSPACE, DOWN_ARROW, ENTER, UP_ARROW} = keyCodes
+
+export default Component.extend({
+  attributeBindings: [
+    'opened:aria-pressed',
+    'role',
+    'style'
+  ],
+
+  classNameBindings: [
+    'disabled:frost-autocomplete-disabled',
+    'error:frost-autocomplete-error',
+    'focused:frost-autocomplete-focused',
+    'opened:frost-autocomplete-opened'
+  ],
+  layout,
+
+  propTypes: {
+    // options
+    isLoading: PropTypes.bool,
+    disabled: PropTypes.bool,
+    error: PropTypes.bool,
+    onChange: PropTypes.func,
+    onClear: PropTypes.func,
+    onFocus: PropTypes.func,
+    onBlur: PropTypes.func,
+    onClick: PropTypes.func,
+    onInput: PropTypes.func,
+    role: PropTypes.string,
+    filterType: PropTypes.oneOf(['startsWith', 'contains']),
+    selectedValue: PropTypes.string,
+    tabIndex: PropTypes.number,
+    width: PropTypes.oneOfType([PropTypes.number, PropTypes.null]),
+    renderTarget: PropTypes.string,
+    filter: PropTypes.string,
+    autofocus: PropTypes.bool,
+    debounceInterval: PropTypes.number,
+
+    // state
+    _element: PropTypes.object,
+    focusedIndex: PropTypes.number,
+    _userInput: PropTypes.bool,
+    _grabfocus: PropTypes.bool,
+    internalSelectedItem: PropTypes.object,
+    opened: PropTypes.bool
+  },
+
+  getDefaultProps () {
+    return {
+      focusedIndex: 0,
+      _userInput: false,
+      _grabfocus: false,
+      filterType: 'startsWith',
+      isLoading: false,
+      disabled: false,
+      autofocus: false,
+      error: false,
+      role: 'button',
+      renderTarget: 'frost-autocomplete',
+      debounceInterval: 0,
+      tabIndex: 0
+    }
+  },
+
+  isFilteredItem (value, filterValue) {
+    if (this.filterType === 'startsWith') {
+      return value.startsWith(filterValue)
+    }
+    return value.indexOf(filterValue) !== -1
+  },
+
+  _handleArrowKey (upArrow) {
```

Comment:
Can you stick with one convention? `_` or without `_`.

Reply:
There's loads of other ones just like this. I'll refrain from adding the same comments to all but assume I mean all of them.